### ISSUE TITLE
Browser CSRF error on REST client in 2.4.137, Security.rest_client_baseurl setting?

### ISSUE THREAD

**Issue by chrisinmtown** [MISP/MISP#6940]
tl;dr: cannot use the web UI's REST client on a version 2.4.137 MISP due to CSRF errors.  This server is behind a reverse proxy server. How can I debug CSRF failures?

We run MISP behind a reverse proxy server that guards the development lab.  This means from the outside I access the MISP UI using a URL like "https://devlab-proxy.mycompany.com:45678", the rev-proxy forwards the traffic to host "https://the-real-machine.mycompany.com:443". 

After a very recent upgrade we are now running MISP 2.4.137 and I was eager to try the new setting Security.rest_client_baseurl. What value did you guys use when you tested this?

To test, I open the page at a URL like https://misp.my-company.com:48752/servers/rest ; pick the templates drop-down, pick events then index; leave the HTTP headers and HTTP body textboxes unchanged from the suggestion; then click the Run query button. I always land at a page "You have tripped the cross-site request forgery protection of MISP." The server's error log is below.
```
2021-03-18 19:52:11 Error: [BadRequestException] The request has been black-holed
Request URL: /servers/rest
Stack Trace:
#0 /var/www/MISP/app/Lib/cakephp/lib/Cake/Controller/Component/SecurityComponent.php(831): AppController->blackhole('secure')
#1 /var/www/MISP/app/Lib/cakephp/lib/Cake/Controller/Component/SecurityComponent.php(351): SecurityComponent->_callback(Object(ServersController), 'blackHole', Array)
#2 /var/www/MISP/app/Lib/cakephp/lib/Cake/Controller/Component/SecurityComponent.php(255): SecurityComponent->blackHole(Object(ServersController), 'secure', Object(SecurityException))
#3 /var/www/MISP/app/Lib/cakephp/lib/Cake/Utility/ObjectCollection.php(129): SecurityComponent->startup(Object(ServersController))
#4 /var/www/MISP/app/Lib/cakephp/lib/Cake/Event/CakeEventManager.php(244): ObjectCollection->trigger('startup')
#5 /var/www/MISP/app/Lib/cakephp/lib/Cake/Controller/Controller.php(683): CakeEventManager->dispatch(Object(CakeEvent))
#6 /var/www/MISP/app/Lib/cakephp/lib/Cake/Routing/Dispatcher.php(189): Controller->startupProcess()
#7 /var/www/MISP/app/Lib/cakephp/lib/Cake/Routing/Dispatcher.php(167): Dispatcher->_invoke(Object(ServersController), Object(CakeRequest))
#8 /var/www/MISP/app/webroot/index.php(104): Dispatcher->dispatch(Object(CakeRequest), Object(CakeResponse))
#9 {main}
```

I tried these values for the Security.rest_client_baseurl setting:

- empty - result is "cross-site request forgery protection"
- the recommend value "https://127.0.0.1" - same result
- the real name of our host, basically "https://host.name.from.etc.hosts" - same result

I don't even know where/when the error is thrown. I don't know if my form with REST client parameters is accepted or not; in other words, do I even reach the second step where the rest_client_baseurl setting is required so the server's PHP page can open a connection to itself?  

To put it mildly I'm super frustrated with seeing this page:
```
You have tripped the cross-site request forgery protection of MISP

CSRF error: This happens usually when you try to resubmit the same form with invalidated CSRF tokens or you had a form open too long and the CSRF tokens simply expired. Just go back to the previous page and refresh the form (by reloading the same url) so that the tokens get refreshed.

Alternatively, click here to continue to the start page.
```
I tend to think the CSRF/XSRF failure is a bug.

Please advise what to try next.

### Work environment

| Questions                 | Answers
|---------------------------|--------------------
| Type of issue             | Bug
| OS version (server)       | RedHat
| OS version (client)       | n/a
| PHP version               | 7.4
| MISP version / git hash   | 2.4.137
| Browser                   | If applicable

### Expected behavior

The web UI's REST client feature functions as designed.

### Actual behavior

The web UI's REST client is inoperable, never get a result

### Steps to reproduce the behavior

Place MISP behing a reverse proxy server.  Set any value for property Security.rest_client_baseurl.  

### Logs, screenshots, configuration dump, ...

I pasted the errors above.

**Comment by chrisinmtown** [comment]
hi CIRCL team, would you please tell me what detail I might provide on CSRF failures?  I just don't know how to make headway here.

In version 2.4.133 we don't see CSRF/XSRF failures.  So I think the XSRF failure is new to version 2.4.137, but I'm not 100% certain.

**Comment by iglocska** [comment]
Hmm, I am not aware of any changes, but I might have missed something there. The form submission / CSRF checks are exactly the same as before, the only change via the new setting is the internal query that MISP executes.

Basically, you post a request to the rest client, rest client does a post request to itself via the rest baseurl, parses the result, returns it to you as the response to the initial request.

Generally CSRF issues will most likely have more to do with your misp.baseurl setting rather than the rest_baseurl.

**Comment by chrisinmtown** [comment]
Thanks Andras @iglocska the baseurl setting looks right to me, it's the outside-the-proxy-server address so all pages generated for users have the right links.  

I'd really like to get the first question answered: does the server declare CSRF upon receipt of the first REST client form POST action, or does that first step succeed?  Any suggestions for debugging that?

**Comment by iglocska** [comment]
The first step fails, the second step has no CSRF protection.

**Comment by chrisinmtown** [comment]
In that case this REST client interaction does not really get started, so I guess the rest_client_baseurl has absolutely no influence.

But I'm baffled that this endpoint declares CSRF when every other endpoint works fine!  I can list events, logs, feeds, blah blah blah.  What do you suggest I investigate here?

Wait.  I should compare apples with apples -- in other words I should enter data on a web-UI form and submit it.  Getting the list of events, list of feeds etc. does not require POST-ing contents of a form.

First test - create a new event.  Put some data in the fields, hit submit -- it works, a new event is created.

But hmm!  When I click the link "Delete event" in the left-side of the page showing the newly created event, I'm back to the "cross-site request forgery protection" error!  Possibly we have a general CSRF problem?  I'm baffled tho that some forms work (like create event), and others do not (delete event).

**Comment by iglocska** [comment]
I must say, that is indeed puzzling\*. Just had a look a look, but can't really explain why just yet. The form is using relative paths to declare the target URL (this is used as part of the mix that goes into generating the CSRF tokens).

\* caveat is that it's approaching midnight over here and my brain is working at half capacity

**Comment by iglocska** [comment]
Stupid question, when you're on the rest client's page, can you just click your URL bar and hit enter? This will generate a fresh set of tokens, just to avoid that your browser is stuck reusing already consumed ones.

**Comment by chrisinmtown** [comment]
I've been force-refreshing the browser by holding shift and clicking the spinny-arrow in Firefox.  I also tried your method, put the cursor at the end of the URL then hit Enter.  The page reloads fine, then I click "Run query" and I'm back to CSRF page.

I'm looking at the request in the Browser debugger. What data field is the CSRF token - maybe data[_Token][key] or data[_Token][fields] ? Can I decode one of those?

**Comment by iglocska** [comment]
data[_Token][key], data[_Token][fields], data[_Token][unlocked] and if enabled data[_Token][debug] all play a role.

Decoding it - don't think so, the important one in this case, key, is a hash generated from a series of variables. Components include a single used shared secret, the target URL to post to and afaik the list of unlocked fields / hash of the list of fields included.

**Comment by iglocska** [comment]
If you're willing to get a bit hacky, could you change `/var/www/MISP/app/View/Servers/rest.ctp` on line 46 from

`echo $this->Form->create('Server', array('novalidate' => true));`

to 

`echo $this->Form->create('Server', array('url' => $baseurl . '/servers/rest', 'novalidate' => true));`

and see if it's any better?

**Comment by chrisinmtown** [comment]
Happy to get a bit hacky :) changed the ctp file (what is that anyhow) but the result is unchanged, upon submit of the REST client form I still get "tripped the cross-site.." page and the error.log file shows the same black-holed error.

Seems to me I need some kind of verbose mode on the server: it should log what it got, what it expected and why it rejected the form POST due to apparent CSRF violation.  But I guess that's not implemented today.

But I wonder if it might be easier, simpler or ? to focus on the delete-event case instead of the complex REST client page: add event, enter event info, click Submit: view the new event; click the Delete event link on the left; in the dialog box "Are you sure you want to delete event #12345" click OK.  For me, bang: CSRF error, and 100% reproducible.

Thanks for the help today/tonite @iglocska maybe we can take this up in the daytime again when we're both thinking straight.

**Comment by iglocska** [comment]
sadly the framework doesn't work that way, debugging CSRF is tricky as that functionality has zero debugging baked in. Both the rest client and the delete event case are inherently just as complex, but we can focus on that, at least it's fewer lines of code that we need to brutalize with some debug calls ;)

I'll try to have a look into what could go wrong today with some ideas how to debug it.

**Comment by chrisinmtown** [comment]
Thanks, please send any ideas you have and I'll immediately try them.  Unfortunately PHP is not one of my core skills.

**Comment by chrisinmtown** [comment]
I believe this will be repaired when https://github.com/MISP/MISP/issues/7329 is repaired, but for now I'm leaving it open.
